crates/router/src/services/api.rs

```diff
@@ -704,32 +704,47 @@ pub enum AuthFlow {
 #[instrument(skip(request, payload, state, func, api_auth), fields(merchant_id))]
 pub async fn server_wrap_util<'a, 'b, A, U, T, Q, F, Fut, E, OErr>(
     flow: &'a impl router_env::types::FlowMetric,
-    state: &'b A,
+    state: web::Data<A>,
     request: &'a HttpRequest,
     payload: T,
     func: F,
     api_auth: &dyn auth::AuthenticateAndFetch<U, A>,
 ) -> CustomResult<ApplicationResponse<Q>, OErr>
 where
-    F: Fn(&'b A, U, T) -> Fut,
+    F: Fn(A, U, T) -> Fut,
     'b: 'a,
     Fut: Future<Output = CustomResult<ApplicationResponse<Q>, E>>,
     Q: Serialize + Debug + 'a,
     T: Debug,
-    A: AppStateInfo,
+    A: AppStateInfo + Clone,
     U: auth::AuthInfo,
     CustomResult<ApplicationResponse<Q>, E>: ReportSwitchExt<ApplicationResponse<Q>, OErr>,
     CustomResult<U, errors::ApiErrorResponse>: ReportSwitchExt<U, OErr>,
     OErr: ResponseError + Sync + Send + 'static,
 {
+    let request_id = RequestId::extract(request)
+        .await
+        .ok()
+        .map(|id| id.as_hyphenated().to_string());
+
+    let mut request_state = state.get_ref().clone();
+
+    request_state.add_request_id(request_id);
+
     let auth_out = api_auth
-        .authenticate_and_fetch(request.headers(), state)
+        .authenticate_and_fetch(request.headers(), &request_state)
         .await
         .switch()?;
+
     let merchant_id = auth_out.get_merchant_id().unwrap_or("").to_string();
+
+    request_state.add_merchant_id(Some(merchant_id.clone()));
```

Comment:
pass the merchant_id instead

Reply:
It is better to pass around None instead of `""`. @kritikmodi please use merchant_id directly and for the log below you can convert it to `""` . It would actually be better to use some const like `"UN_KNOWN"`

Reply:
Okay

Reply:
Updated !